src/redux/actions/dataSet/__tests__/asyncThunks.test.js

```diff
@@ -0,0 +1,81 @@
+import { loadDimensions } from '../asyncThunks'
+import {
+    loadingDimensionsStart,
+    loadingDimensionsSuccess,
+    loadingDimensionsError,
+} from '../../dataSet'
+import { getDimensions } from '../../../../utils/api'
+
+jest.mock('../../dataSet', () => ({
+    loadingDimensionsStart: jest.fn(() => 'loadingDimensionsStart'),
+    loadingDimensionsSuccess: jest.fn(() => 'loadingDimensionsSuccess'),
+    loadingDimensionsError: jest.fn(() => 'loadingDimensionsError'),
+}))
+
+jest.mock('../../../../utils/api', () => ({
+    getDimensions: jest.fn(() => Promise.resolve()),
+}))
+
+describe('Actions - dataSet - asyncThunks', () => {
+    const dispatch = jest.fn()
+
+    describe('loading dimensions', () => {
+        const getState = jest.fn(() => ({
+            dataSet: { selected: {} },
+        }))
+
+        afterEach(() => {
+            loadingDimensionsStart.mockClear()
+            loadingDimensionsSuccess.mockClear()
+            loadingDimensionsError.mockClear()
+        })
+
+        describe('when loading dimensions', () => {
+            it('should dispatch a start loading action', () => {
+                loadDimensions()(dispatch, getState)
+                expect(loadingDimensionsStart).toHaveBeenCalledTimes(1)
+                expect(dispatch).toHaveBeenCalledWith('loadingDimensionsStart')
+            })
+        })
+
+        describe('when loading dimensions successfully', () => {
+            const dimensions = [1, 2, 3, 4]
+
+            it('should dispatch a success action with the dimensions', done => {
+                getDimensions.mockImplementationOnce(() =>
+                    Promise.resolve({ dimensions })
+                )
+
+                loadDimensions()(dispatch, getState).then(() => {
+                    expect(loadingDimensionsSuccess).toHaveBeenCalledTimes(1)
+                    expect(loadingDimensionsSuccess).toHaveBeenCalledWith(
+                        dimensions
+                    )
+                    expect(dispatch).toHaveBeenCalledWith(
+                        'loadingDimensionsSuccess'
+                    )
+                    done()
+                })
+            })
+        })
+
+        describe('when loading dimensions unsuccessfully', () => {
+            const error = new Error('Custom error')
+
+            it('should dispatch an error action with the error', done => {
+                getDimensions.mockImplementationOnce(() =>
+                    Promise.reject(error)
+                )
+
+                loadDimensions()(dispatch, getState).then(() => {
+                    expect(loadingDimensionsError).toHaveBeenCalledTimes(1)
+                    expect(loadingDimensionsError).toHaveBeenCalledWith(error)
+                    expect(dispatch).toHaveBeenCalledWith(
+                        'loadingDimensionsError'
+                    )
+                    done()
+                })
+            })
+        })
+    })
+})
```

Comment:
**First my reasons why I went this way:**
1. I generally think when testing units that are a composition of sub units, it's duplicate effort to test whether the sub units return the right stuff and that it should be tested in a separate test for the sub units.
1. I also wrote the tests for the thunks SUPER quick if I'm not testing the actual actions being dispatched (as the thunk doesn't really care about that, I'm more or less interested if the right action creators have been called and their return values have been dispatched).
1. It guarantees that the thunk is working, especially if we omit tests for the action creators themselves as we're expecting them to work (which means that the correct actions will be dispatched is the correct creators are called)
1. We'll have to test for a subset of actions that are being dispatched anyways, because the actions being dispatched from the tested thunk might be thunks themselves and dispatch multiple actions as well, which will 1. increase the amount of work put into creating the test and 2. result in a higher effort when refactoring the actions that are being dispatched. Here's an example: When loading options for a dropdown and it fails, the error action creator is a thunk itself, dispatching also a feedback action. And testing this behavior is definitely not part of the test for the thunk that actually does the ajax request. It should be tested separately in a test for the error action creator.

**Having said that, I see why this might be the wrong approach for the following reasons:**
1. I'm not following the convention on how to test thunks which might seem weird to other devs
1. It will create inconsistency between apps

I still think that not testing the dispatched actions is a better approach as it's quicker to do and requires less test refactoring when working on the actions themselves, but I'd agree to rewrite them for the two mentioned reasons

Reply:
Thanks for explaining and agreeing to rewrite.
I think it is important to keep testing the actual returned action objects, because essentially the form the contract between the actionCreators and the reducers. Ultimately this is what we want to know: are we dispatching the correct actions to the reducers?